### PR TITLE
COMPONENT_INFORMATION extension and rework

### PR DESCRIPTION
**Pull request by bkueng**
From https://github.com/mavlink/mavlink/issues/1588

- move url to json file (except for the main one + external components, which are always mavlink ftp), allows to use longer URL's
- Standardize the uid to ([MAVLink's](https://mavlink.io/en/services/ftp.html#crc32-implementation)) CRC32. If compression is used, it should be over the compressed files. This has the small advantage of not having to decompress to generate it. It just means that enabling/disabling compression leads to a changed CRC and re-download, but I don't think that's an issue.
- adds a separate optional json file for external components (which is dynamically generated on bootup)
  - each can specify a list of `COMP_METADATA_TYPE`'s with uri's
  - it is explicitly not specified how to e.g. retrieve parameters, just the metadata. This could be done by adding an ID to each component in the metadata.

@DonLakeFlyer can you cross-check this, if it's what you had in mind?

### CODE REVIEW

File: message_definitions/v1.0/common.xml [hunk 1]
```diff
@@ -6646,13 +6649,12 @@
...
-      <field type="char[70]" name="translation_uri">The translations for strings within the metadata file. If null the either do not exist or may be included in the metadata file itself. The translations specified here supercede any which may be in the metadata file itself. The uri format is the same as component_metadata_uri . Files are in Json Translation spec format. Empty string indicates no tranlsation file.</field>
+      <field type="uint32_t" name="general_metadata_file_crc">CRC32 of the TYPE_GENERAL file (can be used by a GCS for file caching).</field>
+      <field type="char[100]" name="general_metadata_uri">Component definition URI for TYPE_GENERAL. This must be a MAVLink FTP URI and the file might be compressed with xz.</field>
+      <field type="uint32_t" name="ext_comp_metadata_file_crc">CRC32 of the TYPE_EXT_COMPONENTS file (can be used by a GCS for file caching).</field>
+      <field type="char[100]" name="ext_comp_metadata_uri">(Optional) Component definition URI for TYPE_EXT_COMPONENTS. This must be a MAVLink FTP URI and the file might be compressed with xz.</field>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml:6657]: Why is this separate? My thinking was that there is just one URI which is used to request all other information about the component. The metadata that is returned would include the connected external types.

**Comment by bkueng** [comment]
As discussed in the call:
- added a fallback URI
- make use of MAVLink's CRC32 used in mavftp (https://mavlink.io/en/services/ftp.html#crc32-implementation)

File: message_definitions/v1.0/common.xml (hunk 1 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml:6657]: @bkueng Your answer in devcall was "because the data is not known at build time, so cannot be supplied in the normal file".

Thinking about this, that's a reason why it can't be returned in the file, not a reason for a new URI. 

Would we be better to revert to specifying the COMP_METADATA_TYPE in the request? It could be either general (default) or COMP_METADATA_TYPE_EXT. It could even be the others too, but then we'd need to decide what to do if one of those was sent (i.e. send the URL and hash for the file still, or just ignore).

**Comment by olliw42** [comment]
couldn't it maybe make sense to change language and to talk of external peripherals or just pheripherals instead of external components?
talking of components in a MAVLink context somehow triggers thoughts it has anything to do, even if only remotely, with MAVLink components...

**Comment by bkueng** [comment]
> couldn't it maybe make sense to change language and to talk of external peripherals or just pheripherals instead of external components?

Yes that's a good point. Changed to peripherals.

**Review by hamishwillee — APPROVED** [review]
Works for me. The point of prototyping this is so that we can test whether we've seen the edge cases.

**Comment by hamishwillee** [comment]
Merged. PS @bkueng Can you keep me informed of progress on implementation. I will need to update the associated docs, but I am assuming this might change a bit more going forward.

**Comment by bkueng** [comment]
Thanks for merging, sure.
